docs(architecture): create Session actor doc (#95)
witnet/witnet-rust: docs/architecture/session.md

# Session

__Session__ is the actor that encapsulates the logic of the sessions 

> associated to each TCP connections.

Reviewer: - This actor actually encapsulates the entire business logic of the Witnet network protocol.
- This actor shouldn't care about the specific transport used by the `ConnectionsManager`, ans so shouldn't its documentation.

## Actor creation and registration

> The creation of the session actor and its registration into the system registry are

Reviewer 2: The creation of the session actor and its addition to the system registry are
performed by the `connection_manager` process:

performed by the `connection_manager` process:

```rust
Session::create(move |ctx| {
    // Get peer address
    let address = stream.peer_addr().unwrap();

    // Split TCP stream into read and write parts
    let (r, w) = stream.split();

    // Add stream in session actor from the read part of the tcp stream
    Session::add_stream(FramedRead::new(r, P2PCodec), ctx);

>     // Create the session actor and store in its state the write part of the tcp stream

Reviewer 2: // Create the session actor and store in its state the write part of the TCP stream

    Session::new(address, session_type, FramedWrite::new(w, P2PCodec, ctx))
});
```

## API

### Outgoing messages: Sessions Manager -> Others

These are the messages sent by the connections manager:

| Message               | Destination             | Input type                                        | Output type                       | Description                           |
|-----------------------|-------------------------|---------------------------------------------------|-----------------------------------|---------------------------------------|

> | Register              | SessionsManager         | `SocketAddr, Addr<Session>, SessionType`          | `SessionsResult<()>`              | Request to register a new session     |

Reviewer: ```suggestion
| `Register`              | `SessionsManager`         | `SocketAddr, Addr<Session>, SessionType`          | `SessionsResult<()>`              | Request to register a new session     |
```

> | Unregister            | SessionsManager         | `SocketAddr, SessionType`                         | `SessionsResult<()>`              | Request to unregister a session       |

Reviewer: ```suggestion
| `Unregister`            | `SessionsManager`         | `SocketAddr, SessionType`                         | `SessionsResult<()>`              | Request to unregister a session       |
```

#### Register

> This message is sent to the [`Sessions Manager`][sessions_manager] actor when the session 

Reviewer: ```suggestion
This message is sent to the [`SessionsManager`][sessions_manager] actor when the session 
```

> actor is started to register this session in the hash map of 

Reviewer: This actor shouldn't care about the specific data structure that `SessionsManager` is using internally for storing references to sessions, and so shouldn't its documentation.

[`Sessions Manager`][sessions_manager].

> The return value is a Result to check if the operation has been done.

Reviewer: ```suggestion
The returned value is a `Result` for easy verification of the success of the operation.
```

> For further information, see [`Sessions Manager`][sessions_manager].

Reviewer: ```suggestion
For further information, see [`SessionsManager`][sessions_manager].
```

#### Unregister
 

> This message is sent to the [`Sessions Manager`][sessions_manager] actor when the session

Reviewer: ```suggestion
This message is sent to the [`SessionsManager`][sessions_manager] actor when the session
```

> actor is stopped to unregister this session in the hash map of the 

Reviewer: This actor shouldn't care about the specific data structure that `SessionsManager` is using internally for storing references to sessions, and so shouldn't its documentation.

[`Sessions Manager`][sessions_manager].

> The return value is a Result to check if the operation has been done.

Reviewer: ```suggestion
The returned value is a `Result` for easy verification of the success of the operation.
```

For further information, see [`Sessions Manager`][sessions_manager].

## Further information

> The full source code of the `Session` can be found at [`session.rs`][session].

Reviewer: ```suggestion
The full source code of the `Session` actor can be found at [`session.rs`][session].
```
